### DISCUSSION TITLE
move window to next (new) workspace

### DISCUSSION THREAD

**Discussion by galkk** [i3/i3#5536] [Ideas]
I want the following functionality. Both proposals are more or less of same nature. The goal is to save me from watching which workspaces do I have now.

# Move windows to next workspace  
If I press button, currently active window moves to next existing workspace.

## Example.
I have workspaces [1,3]. I'm currently on [1]. I press button and current window moves to [3]. (optionally switching to that workspace).

# Move window to new workspace
If I press button, currently active window moves to next non-existing workspace. 

## Example
I have workspaces [1,2,3,5]. I'm currently on [1]. I press button and workspace [4] is created and window moved there (optionally switching to that workspace).

There is a chance that i3 already has it, if so, please point me to it.

Otherwise will there be an interest in it, I can take a look into implementing it.

**Comment by okraits** [comment]
On a general note: please read the userguide yourself, instead of asking questions about wether i3 has a certain feature.

For your first request, you can use:

move [window|container] [to] workspace prev|next|current

See https://i3wm.org/docs/userguide.html#_changing_named_workspaces_moving_to_workspaces

For your second request, there are scripts floating around on the net which do that.

**Reply by galkk** [reply]
Thanks. The guide is quite dense and I missed that part.

**Comment by orestisfl** [comment]
I use a script for that,

```python
#!/usr/bin/python
import json
from subprocess import check_output

workspaces = json.loads(check_output(["i3-msg", "-t", "get_workspaces"]))
workspace_nums = [ws["num"] for ws in workspaces]
f = next(ws["num"] for ws in workspaces if ws["focused"])

for next_num in range(f + 1, 50):
    if next_num not in workspace_nums:  # delete "not" for "Move window to next (existing) workspace"
        print(next_num)
        break
```

Use like this:

```
# switch to next available workspace
bindsym $mod+o exec --no-startup-id i3-msg workspace number $($HOME/bin/i3-next-available-workspace.py)
# move window to next available workspace
bindsym $mod+Shift+o exec --no-startup-id i3-msg move window to workspace $($HOME/bin/i3-next-available-workspace.py) && i3-msg workspace number $_
```

**Reply by yuzujr** [reply]
> i3-next-available-workspace

Thanks,that is exactly what i want

**Comment by ghost** [comment]
bindsym $mod+ctrl+Shift+Left move to workspace prev
bindsym $mod+ctrl+Shift+Right move to workspace next


This worked for me.

**Comment by Ocaly** [comment]
i3/scripts/next_empty_workspace:
```
#!/usr/bin/env bash

workspaces=$(i3-msg -t get_workspaces | tr ',' '\n' | grep num | cut -d ':' -f2 | sort -n | uniq);

for i in $workspaces; do
	if [[ ! "$i" =~ ^[0-9]+$ ]]; then
		echo "got workspace number: '$i'" >&2; exit 1; #grepped a weird num somehow.
	fi;
done;

#if user uses `workspace number 0` start from 0, otherwise 1. idk how to do it without for-loop and regex.
for i in $workspaces; do
	if [[ "$i" =~ ^0$ ]]; then
		idx=0; break;
	elif [[ "$i" =~ ^1$ ]]; then
		idx=1; break;
	else exit 1;
	fi;
done;

while IFS= read -r line; do
	if [[ ! line -eq $idx ]];
		then break;
	fi;
	(( idx++ ))
done <<< $workspaces

echo "$idx"; 
```

i3/config:
```
# open new empty workspace
bindsym $mod+n exec --no-startup-id i3-msg workspace $(~/.config/i3/scripts/next_empty_workspace)
# move container to new empty workspace
bindsym $mod+Shift+n exec --no-startup-id i3-msg move container to workspace number $(~/.config/i3/scripts/next_empty_workspace)
```
Make sure to use `bindsym $mod+n workspace number n` and `bindsym $mod+Shift+n move container to workspace number n` where n are digits 0-9.
